flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,63 +525,88 @@ class ExprLowering {
-  genval(const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
-             &con) {
-    // TODO:
-    // - array constant not handled
-    // - derived type constant
+  mlir::Value genScalarLit(
+      const Fortran::evaluate::Scalar<Fortran::evaluate::Type<TC, KIND>>
+          &value) {
     if constexpr (TC == Fortran::lower::IntegerCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value())
-        return genIntegerConstant<KIND>(builder.getContext(), opt->ToInt64());
-      llvm_unreachable("integer constant has no value");
+      return genIntegerConstant<KIND>(builder.getContext(), value.ToInt64());
     } else if constexpr (TC == Fortran::lower::LogicalCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value())
-        return genLogicalConstantAsI1(builder.getContext(), opt->IsTrue());
-      llvm_unreachable("logical constant has no value");
+      return genLogicalConstantAsI1(builder.getContext(), value.IsTrue());
     } else if constexpr (TC == Fortran::lower::RealCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value()) {
-        std::string str = opt.value().DumpHexadecimal();
-        if constexpr (KIND == 2) {
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEhalf(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else if constexpr (KIND == 4) {
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEsingle(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else if constexpr (KIND == 10) {
-          llvm::APFloat floatVal{llvm::APFloatBase::x87DoubleExtended(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else if constexpr (KIND == 16) {
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        } else {
-          // convert everything else to double
-          llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
-          return genRealConstant<KIND>(builder.getContext(), floatVal);
-        }
+      std::string str = value.DumpHexadecimal();
+      if constexpr (KIND == 2) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEhalf(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 4) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEsingle(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 10) {
+        llvm::APFloat floatVal{llvm::APFloatBase::x87DoubleExtended(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 16) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else {
+        // convert everything else to double
```

Comment:
What else is there?  Why isn't this 'unreachable'?

Reply:
MLIR and LLVM do not support the same floating point semantics specifications. Until they do, we have to hack around it.

Reply:
MLIR set this as the precedent.

Reply:
Aside: this code is in the refactoring queue. It should be using the kind map interface, which was only recently plumbed through.